6473 – Update collector configuration to allow multiple services

vasconsaurus opened the pull request:
Updates the collector configuration to deal with receiving and exporting metrics from multiple services.

### Important
**When thinking about adding multiple receivers/exporters, we need to take into consideration:**
- unlike most otel-collector's receivers and exporters, we can't add multiple prometheus receivers by aliasing, e.g.: `prometheus/service_1`, `prometheus/service_2`. We can have only one receiver, which can have multiple jobs.
- unlike when sending traces, we need to specify the dataset when exporting metrics to honeycomb via the otlp exporter.

**The way we dealt with this is by using the routing connector:**
- we have a pipeline with the prometheus receiver, and the connector functioning as the exporter
- we have a second pipeline with the connector functioning as the receiver, and otlp set up as the exporter
- inside the connector we have a table where we use the service name to decide into which pipeline it should go
    - the prometheus receiver sets the service name from the job_name

**If we want to send prometheus metrics from another service, we would need to:**
- add the endpoint variable to the otel-collector configuration in the docker-compose file (local development)
- add a new job to prometheus' scrape_config,
- add another item to the routing connector table
- add another exporter `otlp/service_metrics`, with the correct Dataset (considering we are also sending metrics to Honeycomb)
- add another pipeline `metrics/service` with the new exporter

**Notes**
- This does add some duplication, but seems the most straightforward way to deal with this, at least for now
- There is a tool to dynamically allocate targets for the prometheus receiver, but it is specific to kubernetes, so not useful for us right now
- Regarding the connector and debugging: the default_pipelines is where anything that is not matched is sent. We currently only use it when debugging. To debug the connector, you can also use the debug pipeline (commented out), and use the debug exporter.

**References**
- [Routing Connector](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/connector/routingconnector)
- [Sending metrics to honeycomb](https://docs.honeycomb.io/send-data/opentelemetry/collector/#metrics-and-logs-signals)
- [Prometheus Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver)
- [Prometheus Receiver Configuration](https://github.com/prometheus/prometheus/blob/v2.28.1/docs/configuration/configuration.md#scrape_config)
- [Two Prometheus receivers can't be defined](https://github.com/open-telemetry/opentelemetry-operator/issues/3034)
- [Target Allocator](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver#opentelemetry-operator)

vasconsaurus committed:
Uupdates with enpoint environment variable

I also renamed  to make it more pender specific, to make it obvious
that those are only pender related metrics.

Adding other exporters and pipelines seems pretty straight forward,
just add '/service_name'. But adding a second Prometheus receiver, or making
sure it sends scraped data to the correct pipeline seems more complex.

dmou commented on otel-collector-config.yaml:
```diff
@@ -2,10 +2,10 @@ receivers:
           scrape_interval: 15s
           static_configs:
-            - targets: ["pender:3200"]
+            - targets: ["${env:pender_metrics_endpoint}"]
```
what if the environment variable was just `prometheus_targets` and included the `["..."]` text? :slightly_smiling_face:

vasconsaurus replied:
A question: Let's say the targets are the pender endpoint, and the check-api endpoint. How do we make sure we send them to the correct exporter, dataset?

I'm wondering if we would have separate prometheus configs for each endpoint, or if we could use one for all of them.

Or does that not matter?

dmou replied:
ah right the way we're using datasets here has them as one per service, and metrics in honeycomb requires it to be attached to a dataset. so we will need to have a separate exporter for each service

we could have them all use the same receiver and then just filter per service based on the metric attribute `app` or something like that using a processor (https://opentelemetry.io/docs/collector/configuration/#processors) but it might be simpler to just have completely separate pipelines :unamused:

vasconsaurus replied:
I found this blog post, which I think might be relevant: https://www.honeycomb.io/blog/simplify-opentelemetry-pipelines-headers-setter

For now, do we want to assume completely separate pipelines or not? If complete separate pipelines, do you still want changes to the env var?

dmou replied:
let's assume separate pipelines and keep the env var you have! I think you'll have to change the prometheus receiver name to be `prometheus/pender` or something similarly unique though (https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/README.md#configuring-receivers)

vasconsaurus replied:
I think that does not work for the prometheus receiver: https://github.com/open-telemetry/opentelemetry-operator/issues/3034

dmou replied:
... well then. it looks like we will have to run an otel collector for each service :unamused: or try to get the processor filtering above working 

ew

vasconsaurus replied:
We can set up different jobs from Prometheus, but the issue here would be how to send each to the correct Dataset, right?

Maybe we can take a step back, and just have one main 'check' Dataset, instead of one per service? I just assumed one per service made sense, I guess. Then we could have a more generic approach. If we need different configuration we could set up a second job, if we don't, we can pass the environment variables as you first suggested.

This would make the configuration easier I think, what do you think? Are there any drawbacks?

vasconsaurus committed:
Add routing connector to OTEL config

So we can route a specific service prometheus metrics
to the specific service honeycomb dataset.

Notes:
1. We don't need to send the dataset when sending traces,
but we still do when sending metrics.
2. While we can aliases most otel-collector components, e.g.:
otlp/pender_metrics. The prometheus receiver does not support that,
so we need the router connector to send to the right pipeline.
(There is a target-allocation feature in the prometheus receiver,
but it is specific to kubernetes.)

vasconsaurus commented:
@dmou I was able to get metrics in Honeycomb using the routing connector 🎉 
If you have some time, please review and let me know what you think.

dmou commented:
> @dmou I was able to get metrics in Honeycomb using the routing connector 🎉 If you have some time, please review and let me know what you think.

amazing! could you update the PR description with how we would add an additional service? just to have some record of what resources would need to be duplicated and which ones can remain :slightly_smiling_face:

vasconsaurus commented:
> > @dmou I was able to get metrics in Honeycomb using the routing connector 🎉 If you have some time, please review and let me know what you think.
> 
> amazing! could you update the PR description with how we would add an additional service? just to have some record of what resources would need to be duplicated and which ones can remain 🙂

@dmou, I updated the PR, let me know what you think.

dmou reviewed (approved):
lgtm if you're intentionally keeping the `for debugging purposes` lines for now!
